File: Sample/Engine/UI/ListView.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
from typing import List, Tuple, Union
from .. import Angle, Pair, IntRect, RenderTarget, RenderStates, Transform, TypeAdapter, Vector2f, Vector2i, degrees
from ..Utils import Math, Render
from .Base import ControlBase, FunctionalBase


class ListView(ControlBase, FunctionalBase):
    r"""List view control that arranges child controls in a grid layout.

    Supports multiple columns and automatic item positioning.
    """

    def __init__(
        self,
        rect: Union[IntRect, Tuple[Pair[int], Pair[int]]],
        defaultItemHeight: int = 32,
        fixItemHeight: bool = False,
        columns: int = 1,
    ) -> None:
        r"""\brief Construct a ListView with a given rectangle and layout options.

        - \param rect              Position and size of the list view
        - \param defaultItemHeight  Default height for items (used when item has no size)
        - \param fixItemHeight     Whether to force all items to the same height
        - \param columns           Number of columns in the grid
        """
        super().__init__()
        FunctionalBase.__init__(self)
        self.size = rect.size
        self.setPosition(Math.ToVector2f(rect.position))
        self._defaultItemHeight: int = defaultItemHeight
        self._fixItemHeight: bool = fixItemHeight
        self._columns: int = columns
        self._children: List[ControlBase] = []
        self._renderStates: RenderStates = Render.CanvasRenderStates()
        self._positionsSettled: bool = False

    def v_getPosition(self) -> Pair[float]:
        r"""\brief Get the list view position as a plain pair.

        - \return  Position as (x, y)
        """
        result = super().getPosition()
        return (result.x, result.y)

    @TypeAdapter(position=([tuple, list], Vector2f))
    def setPosition(self, position: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Set the list view position.

        - \param position  New position in logical UI units
        """
        super().setPosition(position)

    @TypeAdapter(offset=([tuple, list], Vector2f))
    def move(self, offset: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Move the list view by an offset.

        - \param offset  Offset to add to the current position
        """
        super().move(offset)

    def v_getRotation(self) -> float:
        r"""\brief Get the list view rotation as a plain number.

        - \return  Rotation in degrees
        """
        result = super().getRotation()
        return result.asDegrees()

    def setRotation(self, angle: Union[Angle, float]) -> None:
        r"""\brief Set the list view rotation.

        - \param angle  Rotation (Angle object or degrees)
        """
        if not isinstance(angle, Angle):
            angle = degrees(angle)
        super().setRotation(angle)

    def rotate(self, angle: Union[Angle, float]) -> None:
        r"""\brief Rotate the list view by an offset.

        - \param angle  Rotation offset (Angle object or degrees)
        """
        if not isinstance(angle, Angle):
            angle = degrees(angle)
        super().rotate(angle)

    def v_getScale(self) -> Pair[float]:
        r"""\brief Get the list view scale as a plain pair.

        - \return  Scale as (x, y)
        """
        result = super().getScale()
        return (result.x, result.y)

    @TypeAdapter(scale=([tuple, list], Vector2f))
    def setScale(self, scale: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Set the list view scale.

        - \param scale  New scale
        """
        super().setScale(scale)

    @TypeAdapter(factor=([tuple, list], Vector2f))
    def scale(self, factor: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Scale the list view by a factor.

        - \param factor  Scale factor to multiply by
        """
        super().scale(factor)

    def v_getOrigin(self) -> Pair[float]:
        r"""\brief Get the list view origin as a plain pair.

        - \return  Origin as (x, y) in logical UI units
        """
        origin = self.getOrigin()
        return (origin.x, origin.y)

    def getOrigin(self) -> Vector2f:
        r"""\brief Get the list view origin in logical UI units.

        - \return  Origin position in logical UI units
        """
        from .. import Scale

        origin = super().getOrigin()
        return origin / Scale

    @TypeAdapter(origin=([tuple, list], Vector2f))
    def setOrigin(self, origin: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Set the list view origin in logical UI units.

        - \param origin  Origin position in logical UI units
        """
        from .. import Scale

        super().setOrigin(origin * Scale)

    def getColumns(self) -> int:
        r"""\brief Get the number of columns in the list view.

        - \return  Current column count
        """
        return self._columns

    def setSize(self, size: Vector2i) -> None:
        r"""\brief Set the list view size and invalidate child layout.

        - \param size  New list view size
        """
        self.size = size
        self._positionsSettled = False

    def setColumns(self, columns: int) -> None:
        r"""\brief Set the number of columns in the list view.

        - \param columns  New column count
        """
        self._columns = columns
        self._positionsSettled = False

    def getChildren(self) -> List[ControlBase]:
        r"""\brief Get the list of child controls.

        - \return  List of child controls
        """
        return self._children

    def addChild(self, child: ControlBase) -> None:
        r"""\brief Add a child control to the list view.

        - \param child  Control to add
        """
        self._children.append(child)
        child.setParent(self)
        self._positionsSettled = False

    def removeChild(self, child: ControlBase) -> None:
        r"""\brief Remove a child control from the list view.

        - \param child  Control to remove
        """
        self._children.remove(child)
        child.setParent(None)
        self._positionsSettled = False

    def clearChildren(self) -> None:
        r"""\brief Remove all child controls from the list view."""
        for child in self._children:
            child.setParent(None)
        self._children.clear()
        self._positionsSettled = False

    def getRenderStates(self) -> RenderStates:
        r"""\brief Get the render states used when drawing this list view.

        - \return  Current render states
        """
        return self._renderStates

    def update(self, deltaTime: float) -> None:
        r"""\brief Update visible children in the list view.

        - \param deltaTime  Time elapsed since last update, in seconds
        """
        for child in self._children:
            if not (isinstance(child, FunctionalBase) and child.getVisible()):
                continue
            child.update(deltaTime)

    def lateUpdate(self, deltaTime: float) -> None:
        r"""\brief Run late update on children in the list view.

        - \param deltaTime  Time elapsed since last update, in seconds
        """
        for child in self._children:
            if not (isinstance(child, FunctionalBase) and child.getVisible()):
                continue
            child.lateUpdate(deltaTime)

    def fixedUpdate(self, fixedDelta: float) -> None:
        r"""\brief Run fixed-timestep update on children in the list view.

        - \param fixedDelta  Fixed timestep duration, in seconds
        """
        for child in self._children:
            if not (isinstance(child, FunctionalBase) and child.getVisible()):
                continue
            child.fixedUpdate(fixedDelta)

    def draw(self, target: RenderTarget, states: RenderStates) -> None:
        r"""\brief Draw the list view and its children to a render target.

        - \param target  Render target used for drawing
        - \param states  Render states used when drawing
        """
        self.applyPositions()
        states.transform *= self.getTransform()
        for child in self._children:
            target.draw(child, states)

    def applyPositions(self) -> None:
        r"""\brief Recompute child positions in the grid layout.

        Automatically called before drawing if positions have not been settled.
        """
        if self._positionsSettled:
            return
        self._positionsSettled = True
        colWidth = (self.size.x - 32) / self._columns
        rowHeight = 0
        currentY = 0
        currentRowHeights = []

        for index, child in enumerate(self._children):
            col = index % self._columns
            row = index // self._columns
            if col == 0 and index > 0:
                currentY += rowHeight if rowHeight > 0 else self._defaultItemHeight
                rowHeight = 0
                currentRowHeights = []
            itemHeight = 0
            itemWidth = 0
            if isinstance(child, ControlBase):
                size = child.getSize()
                itemHeight = size.y
                itemWidth = size.x
            itemHeight = max(itemHeight, self._defaultItemHeight)
            if self._fixItemHeight:
                itemHeight = self._defaultItemHeight
            colCentre = 16 + col * colWidth + colWidth / 2
            originX = child.getOrigin().x
            scaleX = child.getScale().x
            posX = colCentre - itemWidth / 2 + originX * scaleX
            currentRowHeights.append(itemHeight)
            rowHeight = max(rowHeight, itemHeight)
            child.setPosition((posX, currentY))

    def _getRenderTransform(self) -> Transform:
        from .. import Scale

        transform = Transform()
        transform *= self.getTransform()
        transform.translate(self.getPosition() * (Scale - 1))
        return transform

File: Sample/Engine/Source/Core/src/UI/ListView.cpp
#include <UI/ListView.hpp>

#include <EngineState.hpp>
#include <Utils/Render.hpp>

#include <algorithm>
#include <stdexcept>

ListView::ListView(const sf::IntRect& rect, int defaultItemHeight,
                   bool fixItemHeight, int columns)
    : size_(static_cast<float>(rect.size.x), static_cast<float>(rect.size.y)),
      defaultItemHeight_(defaultItemHeight),
      fixItemHeight_(fixItemHeight),
      columns_(columns),
      renderStates_(canvasRenderStates()),
      displayScale_(Scale) {
    if (columns_ <= 0) {
        throw std::invalid_argument("ListView columns must be positive");
    }
    setPosition({static_cast<float>(rect.position.x),
                 static_cast<float>(rect.position.y)});
}

sf::Vector2f ListView::getOrigin() const {
    return ControlBase::getOrigin() / Scale;
}

void ListView::setOrigin(const sf::Vector2f& origin) {
    ControlBase::setOrigin(origin * Scale);
}

int ListView::getColumns() const {
    return columns_;
}

sf::Vector2f ListView::getSize() const {
    return size_;
}

sf::FloatRect ListView::getContentBounds() const {
    const_cast<ListView*>(this)->applyPositions();
    sf::FloatRect bounds{{0.0f, 0.0f}, size_};
    for (const std::shared_ptr<ControlBase>& child : children_) {
        if (child == nullptr || !child->getVisible()) {
            continue;
        }
        const sf::FloatRect childBounds =
            child->getTransform().transformRect(child->getContentBounds());
        const float minimumX =
            std::min(bounds.position.x, childBounds.position.x);
        const float minimumY =
            std::min(bounds.position.y, childBounds.position.y);
        const float maximumX =
            std::max(bounds.position.x + bounds.size.x,
                     childBounds.position.x + childBounds.size.x);
        const float maximumY =
            std::max(bounds.position.y + bounds.size.y,
                     childBounds.position.y + childBounds.size.y);
        bounds = {{minimumX, minimumY},
                  {maximumX - minimumX, maximumY - minimumY}};
    }
    return bounds;
}

void ListView::setSize(const sf::Vector2i& size) {
    setSizeValue({static_cast<float>(size.x), static_cast<float>(size.y)});
}

void ListView::setSizeVector2u(const sf::Vector2u& size) {
    setSizeValue({static_cast<float>(size.x), static_cast<float>(size.y)});
}

void ListView::setSizeVector2f(const sf::Vector2f& size) {
    setSizeValue(size);
}

void ListView::setColumns(int columns) {
    if (columns <= 0) {
        throw std::invalid_argument("ListView columns must be positive");
    }
    columns_ = columns;
    positionsSettled_ = false;
}

std::vector<std::shared_ptr<ControlBase>> ListView::getChildren() const {
    return children_;
}

void ListView::addChild(const std::shared_ptr<ControlBase>& child) {
    if (child == nullptr) {
        throw std::invalid_argument("ListView child cannot be null");
    }
    const std::shared_ptr<ControlBase> self = weak_from_this().lock();
    if (self == nullptr) {
        throw std::logic_error("ListView owner is not shared");
    }
    children_.push_back(child);
    child->setParent(self);
    positionsSettled_ = false;
}

void ListView::removeChild(const std::shared_ptr<ControlBase>& child) {
    const auto iterator = std::find(children_.begin(), children_.end(), child);
    if (iterator == children_.end()) {
        throw std::invalid_argument("ListView child not found");
    }
    (*iterator)->setParent(nullptr);
    children_.erase(iterator);
    positionsSettled_ = false;
}

void ListView::clearChildren() {
    for (const std::shared_ptr<ControlBase>& child : children_) {
        child->setParent(nullptr);
    }
    children_.clear();
    positionsSettled_ = false;
}

sf::RenderStates ListView::getRenderStates() const {
    return renderStates_;
}

void ListView::update(float deltaTime) {
    for (const std::shared_ptr<ControlBase>& child : children_) {
        FunctionalBase* functional = dynamic_cast<FunctionalBase*>(child.get());
        if (functional != nullptr && child->getVisible()) {
            functional->update(deltaTime);
        }
    }
    FunctionalBase::update(deltaTime);
}

void ListView::lateUpdate(float deltaTime) {
    for (const std::shared_ptr<ControlBase>& child : children_) {
        FunctionalBase* functional = dynamic_cast<FunctionalBase*>(child.get());
        if (functional != nullptr && child->getVisible()) {
            functional->lateUpdate(deltaTime);
        }
    }
}

void ListView::fixedUpdate(float fixedDelta) {
    for (const std::shared_ptr<ControlBase>& child : children_) {
        FunctionalBase* functional = dynamic_cast<FunctionalBase*>(child.get());
        if (functional != nullptr && child->getVisible()) {
            functional->fixedUpdate(fixedDelta);
        }
    }
}

void ListView::invalidatePositions() {
    positionsSettled_ = false;
}

void ListView::applyPositions() {
    if (positionsSettled_) {
        return;
    }
    positionsSettled_ = true;
    const float columnWidth = (size_.x - 32.0f) / columns_;
    float rowHeight = 0.0f;
    float currentY = 0.0f;

    for (std::size_t index = 0; index < children_.size(); ++index) {
        const std::shared_ptr<ControlBase>& child = children_[index];
        const int column = static_cast<int>(index % columns_);
        if (column == 0 && index > 0) {
            currentY += rowHeight > 0.0f
                            ? rowHeight
                            : static_cast<float>(defaultItemHeight_);
            rowHeight = 0.0f;
        }
        float itemHeight = child->getSize().y;
        itemHeight =
            std::max(itemHeight, static_cast<float>(defaultItemHeight_));
        if (fixItemHeight_) {
            itemHeight = static_cast<float>(defaultItemHeight_);
        }
        const float columnCentre = 16.0f +
                                   static_cast<float>(column) * columnWidth +
                                   columnWidth / 2.0f;
        const sf::FloatRect bounds = child->getLocalBounds();
        const float originX = child->getOrigin().x;
        const float scaleX = child->getScale().x;
        const float positionX =
            columnCentre -
            (bounds.position.x + bounds.size.x / 2.0f - originX) * scaleX;
        rowHeight = std::max(rowHeight, itemHeight);
        child->setPosition({positionX, currentY});
    }
}

void ListView::refreshDisplayScale() {
    if (displayScale_ != Scale) {
        const sf::Vector2f logicalOrigin =
            ControlBase::getOrigin() / displayScale_;
        displayScale_ = Scale;
        setOrigin(logicalOrigin);
        invalidatePositions();
    }
    ControlBase::refreshDisplayScale();
}

void ListView::draw(sf::RenderTarget& target, sf::RenderStates states) const {
    const_cast<ListView*>(this)->applyPositions();
    states.transform.combine(getTransform());
    for (const std::shared_ptr<ControlBase>& child : children_) {
        target.draw(*child, states);
    }
}

void ListView::setSizeValue(const sf::Vector2f& size) {
    size_ = size;
    positionsSettled_ = false;
}
